Add Gnosis Safe event message

willm30 opened the pull request:
## Description

This PR tracks the work involved with ensuring that the correct messages appear on the Action Event page when a new safe is added, and that the correct `ActionListItem` titles appear in the Colony Actions list. 

Design: https://www.figma.com/file/LlVmeMQxHVA04evfrgZyuN/Safe-Control?node-id=727%3A15054

**Changes** 🏗

* DefaultAction & ActionsListItem: Show 'Add Gnosis Safe' title when safe added
* Details Widget: Add AddSafe view when safe added
* Updating `staticMaps`, the i18n files, and various other files to account for the AddSafe action

## Screenshot

![add-gnosis-safe](https://user-images.githubusercontent.com/64402732/180331605-a81cbb2e-9480-44af-b549-10353059edda.png)

![actionsList](https://user-images.githubusercontent.com/64402732/180656265-35d92abe-06ef-4fd0-9156-0c992f8bb741.png)


## TODO

- [x] Event title says 'Add Gnosis Safe'
- [x] Event text accurately describes the added safe
- [x] Details Widget displays correct data 
- [x] ActionsList on ColonyHome displays correct titles 

Resolves #3657

ArmandoGraterol reviewed (changes requested):
I think something changed in the base PR that broke the add safe action? Seeing it's the same error @danbr pointed out in #3674
![Screenshot from 2022-07-27 18-01-40](https://user-images.githubusercontent.com/18473896/181371611-928bb379-d801-49ac-bee8-d353bfc6d0aa.png)

Left some comments for things I've noticed for now that might need changes

willm30 commented on src/modules/dashboard/components/ActionsPage/ActionsComponents/RecoveryAction.tsx:
```diff
@@ -271,6 +274,7 @@ const RecoveryAction = ({
                       }}
                       emmitedBy={emmitedBy}
                       colony={colony}
+                      colonyMetadataChecks={metadataChecks}
```
@ArmandoGraterol Yep, have removed from both

willm30 commented on src/utils/colonyActions.ts:
```diff
@@ -265,6 +274,58 @@ export const sortMetdataHistory = (colonyMetadata) =>
+    return currentMetadata.metadata;
+  }
+};
+/* eslint-enable consistent-return */
```
Thank you for explaining. Hadn't come across that before and was confused!

willm30 commented:
> I think something changed in the base PR that broke the add safe action? 
> 
> Left some comments for things I've noticed for now that might need changes

@ArmandoGraterol I've fixed this on the base branch and have rebased the others so this should be fixed. Thanks for explaining why the linter wasn't happy with me. I've implemented your other suggestions.

ArmandoGraterol commented on src/modules/core/components/ActionsList/ActionsListItem.tsx:
```diff
@@ -287,7 +288,7 @@ const ActionsListItem = ({
                 id={roleMessageDescriptorId || 'action.title'}
                 values={{
-                  actionType,
+                  actionType: isActionTypeAddSafe ? 'AddSafe' : actionType,
```
Are there any reasons why this type cannot be determined in `getActionType` like the types of the other actions? This isn't the first action that uses the same events or it's quite similar to other actions, so I'm wondering why can this be done  in the resolvers when others in a similar context can

willm30 replied:
@ArmandoGraterol I think originally I struggled to see where I'd implement the checks, but after looking again I've seen where they should go and have included them in the transformer (for the Colony Actions list) and the resolver (for the Actions Page).

ArmandoGraterol commented on src/modules/dashboard/components/ActionsPage/ActionsComponents/DefaultAction.tsx:
```diff
@@ -178,10 +201,18 @@ const DefaultAction = ({
 
   const actionAndEventValuesForDocumentTitle = {
-    actionType,
+    actionType: isActionTypeAddSafe ? 'AddSafe' : actionType,
```
Same question here

ArmandoGraterol reviewed (commented):
Looking great from what I've been testing! Congrats on your first action page wiring :) The IPFS workaround of passing the data in the router state, if I understand it correctly, seems like a good idea, haven't thought about doing something like that before.

I just have a question regarding determining the action type on the action page.

ArmandoGraterol commented on src/modules/dashboard/components/ActionsPage/DetailsWidget/DetailsWidget.tsx:
```diff
@@ -153,6 +165,34 @@ const DetailsWidget = ({
           </div>
         </div>
       )}
+      {values?.isAddSafe && (
```
Also, you need to change.  Add the details of this action and the action type itself to the `DETAILS_FOR_ACTION` mapping that we have, that's something that we do for every action. Then each detail should have its own separate conditional like the rest of the details in this component:

```
...
{detailsForAction.Chain && (...)}
{detailsForAction.Safe && (...)}
...
```

This is really useful and important because the component stays adaptable to any other action that might share details with this one.

ArmandoGraterol replied:
Overall, take a look at the `staticMaps` file and make sure that this new action type is correctly integrated just like the other action types that we have

willm30 commented:
@ArmandoGraterol I've added the checks for whether the action type is "AddSafe" to `getActionType` in the resolver (in the case of the Actions Page) and the `getActionsListData` transformer (for the actions list on the colony home page), as you suggested. I've also updated the StaticMaps so the DetailsWidget follows the same pattern as the other actions.

Julianahlk reviewed (changes requested):
Great work so far Will! However, I am experiencing the following:

1. Whenever I add a new safe I see the following data displayed in the DetailsWidget: 
<img width="1514" alt="Screen Shot 2022-08-02 at 6 10 14 PM" src="https://user-images.githubusercontent.com/71563622/182424450-5f995c11-9bcf-4cc4-9a2d-00b30754f6b2.png">
But I was expecting to see more data about the safe like shown below:
<img width="1005" alt="Screen Shot 2022-08-02 at 6 12 13 PM" src="https://user-images.githubusercontent.com/71563622/182425420-5490ea22-199d-46f8-99b1-3e1116d34a32.png">

2. Every time I add a new safe I see two action list items being created but I would expect just one to be created:
<img width="991" alt="Screen Shot 2022-08-02 at 6 12 44 PM" src="https://user-images.githubusercontent.com/71563622/182427231-5da16bfe-31f5-40e0-8c7d-c9f125c84d36.png">

- The first item from the bottom up takes me to the page below and it contains all the expected data details of the newly added safe:
<img width="991" alt="Screen Shot 2022-08-02 at 6 12 44 PM" src="https://user-images.githubusercontent.com/71563622/182427476-a100851e-5f4c-465f-8a70-a91bbe46f9c8.png">

- The second item takes me to a blank page with some TypeErrors:
<img width="1611" alt="Screen Shot 2022-08-02 at 6 38 19 PM" src="https://user-images.githubusercontent.com/71563622/182427788-875fa550-a9d5-4c0c-8df5-e3c71e698ff9.png">

willm30 commented:
> 1. Whenever I add a new safe I see the following data displayed in the DetailsWidget:

This should now be fixed, thanks.

> 2. Every time I add a new safe I see two action list items being created but I would expect just one to be created:

This doesn't happen on my machine. Does this still happen after restarting webpack or after a creating a new colony? And if so, does this happen for all actions or just add safe? Could you let me know anything relevant for me to reproduce, e.g. which browser are you using, which extensions have you got installed? Thanks.

Julianahlk commented:
> This doesn't happen on my machine. Does this still happen after restarting webpack or after a creating a new colony? And if so, does this happen for all actions or just add safe? Could you let me know anything relevant for me to reproduce, e.g. which browser are you using, which extensions have you got installed? Thanks.

1. Yes, still experiencing issues even after restarting Webpack. 
2. I tried creating a new Colony but for some reason I am unable to do so. 
3. At the moment, every time I create a new action, the actions list is not updating for all actions.
4. 
<img width="809" alt="Screen Shot 2022-08-02 at 8 10 25 PM" src="https://user-images.githubusercontent.com/71563622/182444392-e94fc495-3403-45ad-8068-b287d8530093.png">

Julianahlk commented:
Hey Will, I re-started my dev env and things are working now. I sent you some videos over Discord of me creating a safe for the first time to make sure everything looks right.

chinins reviewed (changes requested):
Nice attempt but unfortunately not a correct one.

The approach that you are using is not in line with what we already have in the dapp and also duplicates some of the already existing logic. It's not your fault - you don't know about some things in the dapp. The feature branch is not rebased on latest master where verified recipients is introduced. There, the logic for comparing colony metadata already exists.
I think that you should wait for Armando to rebase on latest master. Then rebase your branch and then update this PR based on the metadata checks that we already have in the dapp and the approach we use.

I've already discussed with Armando that for this issue & for the issue that I have of adding remove safe page we need the rebase and this is on his radar.

Also, for the PR titles - these are not commits, no need to add `Add:` (Feat or anything like this). It seems that this PR is to add safe action page so just `Add sage action page` would be good.

willm30 commented:
@chinins I have updated to reflect your comment and the existing style of checking metadata in the dapp. I'd be grateful if you couuld take another look to confirm this is what you were thinking. 

As an aside, I think we should update all the sagas that interact with colony metadata to use the approach in the verified recipients saga. Do you and @ArmandoGraterol agree with that? When I was working on wiring up the addSafe saga, I changed the editColony saga to get the current metadata direct from IPFS and update it that way, but verified recipients receives all the metadata via its payload. So, I think we should ensure consistency. They also all need to be updated to ensure they're all including the safe data -- I haven't done this is in this branch but would imagine including that in the work to ensure consistency across them all.

chinins reviewed (changes requested):
Nice work! I think that you've chosen an approach that is too manual so prone to errors if there is added complexity and it also introduces yet another way of handling actions.

I was able to get into nitty-gritty details because I was working on a similar PR - just removing safes. Please, have a look at my PR #3736 . I tried to generalise and abstract the colony edit actions specifics. The approach for adding and removing safes should be the same so if you and the rest of the team is happy with my approach then please rebase your branch on mine and refactor accordingly. Alternatively I'm open to suggestions.

For the sagas/actions I think it should all be handled in a separate issue/PR. I talked to @ArmandoGraterol  a while ago about refactoring them after a rebase and I think it should be on his radar to add an issue for that. I think it makes sense to combine it all in one PR - to make sure that it's exactly the same approach everywhere. If @ArmandoGraterol  agrees with me I would remove that part and let it be handled separately. But Armando is the feature owner so it's for him to decide.

willm30 commented:
@chinins I have updated to follow your approach. Please review when you get a chance. Thanks

willm30 commented on src/utils/colonyActions.ts:
```diff
@@ -323,13 +337,14 @@ export const getSpecificActionValuesCheck = (
       const nameChanged = prevColonyDisplayName !== currentColonyDisplayName;
       const logoChanged = prevColonyAvatarHash !== currentColonyAvatarHash;
-
+      const safeAdded = currentColonySafes.length > prevColonySafes.length;
```
@chinins For this one, I updated the `colonyAction` resolver so that if `colonySafes` comes back from IPFS as `null` it will return an empty array. This is consistent with how we've typed the colonySafes property (i.e. as non-nullable).

ArmandoGraterol reviewed (approved):
Looking good to me! Tested by adding several safes and it was working as expected. The code is much simpler than the first I reviewed it :)

About the sagas, yeah it was agreed that after the rebase updating the sagas added in this feature would also be updated with a dedicated issue/PR

chinins commented on src/i18n/en-actions.ts:
```diff
@@ -33,6 +33,7 @@ const actionsMessageDescriptors = {
       ${ColonyExtendedActions.AddressBookUpdated} {Address book was updated}
       ${ColonyExtendedActions.TokensUpdated} {Colony tokens were updated}
       ${ColonyExtendedActions.SafeRemoved} {Remove Safe}
+      ${ColonyExtendedActions.SafeAdded} {Add Gnosis Safe}
```
We might have a different chain than Gnosis so either you remove Gnosis from the string or you add a variable and pass the relevant chain name.

arrenv replied:
This would be: "Add Safe From Rinkeby Testnet"

chinins commented on src/i18n/en-events.ts:
```diff
@@ -30,6 +30,7 @@ const eventsMessageDescriptors = {
   [`event.${ColonyAndExtensionsEvents.ColonyMetadata}.tokens`]: `{initiator} changed this colony's tokens`,
   [`event.${ColonyAndExtensionsEvents.ColonyMetadata}.verifiedAddresses`]: `{initiator} updated this colony's address book`,
   [`event.${ColonyAndExtensionsEvents.ColonyMetadata}.safeRemoved`]: `{initiator} removed {removedSafesString} using the Administration and Funding permissions.`,
+  [`event.${ColonyAndExtensionsEvents.ColonyMetadata}.safeAdded`]: `{initiator} added the Gnosis Safe {safeAddress} from {selectedChain} using the Administration and Funding permissions`,
```
Same here - we might have a safe from a different chain so you need to add the chainName variable and pass it in (same as safeAddress).
@ArmandoGraterol, am I right?

ArmandoGraterol replied:
This might be a good question for @arrenv as from my understanding these safes may still be called "Gnosis Safe" even though they may not be in the "Gnosis Chain". 

For example, in `https://gnosis-safe.io` you can create safes from different chains other than the Gnosis chain, my suspicion is that "Gnosis Safe" may just be a "brand name" for the safes.

willm30 replied:
I understand it that the product is called "Gnosis Safe", and the use of the term "safe" to refer to a multisig is specific to the Gnosis Safe platform. Therefore something like an "Aurora" safe doesn't exist. It would be a Gnosis Safe deployed on the Aurora network. 

@arrenv Is that right?

arrenv replied:
@willm30 @ArmandoGraterol  Yes, "Gnosis Safe" is a brand. In fact they are rebranding it to just be "Safe" - https://www.cryptotimes.io/gnosis-safe-is-rebranded-to-safe-with-100m-fundraise

Yes, we need to be able to specify what chain the Safe is on.

arrenv replied:
So, this would be: "{initiator} added the Safe {safeAddress} from {selectedChain} using the Administration and Funding permissions`"

chinins commented on src/modules/dashboard/components/ActionsPage/ActionsComponents/DefaultAction.tsx:
```diff
@@ -231,10 +231,15 @@ const DefaultAction = ({
     isSmiteAction: new Decimal(reputationChange).isNegative(),
     removedSafes,
     removedSafesString,
+    safeAddress: addedSafe && (
```
It's a nitpick but I would call it something like `addedSafeAddress`, so it's clear what operation we are talking about. But up to you...

chinins commented on src/modules/dashboard/components/ActionsPage/ActionsComponents/DefaultAction.tsx:
```diff
@@ -231,10 +231,15 @@ const DefaultAction = ({
+    safeAddress: addedSafe && (
+      <MaskedAddress address={addedSafe.contractAddress} />
+    ),
+    selectedChain: addedSafe && GNOSIS_SAFE_NAMES_MAP[addedSafe.chainId],
```
A nitpick again but I would go for a more precise name, something like `chainName` (name is an important part here). But up to you...

chinins commented on src/utils/colonyActions.ts:
```diff
@@ -364,6 +368,14 @@ export const getSpecificActionValuesCheck = (
+        currentColonySafes.length > prevColonySafes.length
+          ? currentColonySafes.filter(
+              (safe) =>
+                !prevColonySafes.some((prevSafe) => isEqual(prevSafe, safe)),
```
I think you can just do a simple check of `prevSafe.address === safe.address`.
I think that you can also use `.find` instead of filter, this would give you the element without the need to select `[0]` like you do now. But that's a nitpick.

chinins reviewed (approved):
Looks much better and cleaner - good job!

There is only one thing that I also mentioned in my direct comments. The safe name in the title and description should be based on the chain name or removed at all (at least htis is how I understand the feature) but it's hardcoded. So what ends up happening is I add a safe from Aurora but it's called Gnosis. That doesn't seem right. @ArmandoGraterol, @arrenv, can you please confirm?
Unrelated to this issue, but if I'm right I would also suggest to update the title in the advanced dialog from `Gnosis Safe Control` to just `Safe control`.

<img width="1034" alt="Screenshot 2022-08-17 at 11 31 55" src="https://user-images.githubusercontent.com/34057551/185078189-e6b7744c-9de3-4ae9-8878-16f6a4ee44c2.png">

arrenv reviewed (approved):
Looks and works well! Very nice!

To add to @chinins comments, I also think we should improve this to use "Safe" as the naming and then include the chainName.
